### PR TITLE
Cscoiva 972: Attachments for Kuljettajakoulutukset

### PR DESCRIPTION
**Pull request by sz1521**
Refactored attachment file getting function etc.

### CODE REVIEW

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Hakemus/hakemusContainer.js [hunk 1]
```diff
@@ -185,6 +185,11 @@ const HakemusContainer = props => {
         return changeObjects;
       };
 
+      let perustelutLiiitteetChanges =
+        R.path(
+          ["meta", "tutkinnotjakoulutuksetLiitteet", "changeObjects"],
+          muutospyynnot.muutospyynto
+        ) || [];
```
Comment (rapitkan) [inline src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Hakemus/hakemusContainer.js]: Variable name could be more specific. The code is fetching the attachments of the Tutkinnot ja koulutukset section but the variable (perustelutLiitteetChanges) doesn't indicate it.

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js [hunk 2]
```diff
@@ -222,33 +237,36 @@ const MuutospyyntoWizard = props => {
...
-    const allAttachments = (
-      R.path(["yhteenveto", "yleisettiedot"], changeObjects) || []
-    ).concat(
-      (R.path(["yhteenveto", "yleisettiedot"], changeObjects) || []).concat(
-        (
-          R.path(["yhteenveto", "hakemuksenliitteet"], changeObjects) || []
-        ).concat(R.path(["taloudelliset", "liitteet"], changeObjects) || [])
+    const allAttachments = combineArrays([
+      R.path(["yhteenveto", "yleisettiedot"], changeObjects) || [],
+      R.path(["yhteenveto", "yleisettiedot"], changeObjects) || [],
```
Comment (rapitkan) [inline src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js]: Why the same code is here twice?

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js [hunk 3]
```diff
@@ -222,33 +237,36 @@ const MuutospyyntoWizard = props => {
...
-          return R.map(file => {
-            return file;
+      R.forEachObjIndexed(obj => {
+        if (obj.properties.attachments) {
+          R.forEachObjIndexed(file => {
+            attachments.push(file);
           }, obj.properties.attachments);
-        else return null;
+        }
       }, allAttachments);
```
Comment (rapitkan) [inline src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js]: R.filter can be used here for cleaner implementation.

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardPerustelut.js [hunk 4]
```diff
@@ -19,10 +19,14 @@ import {
...
 import FormSection from "../../../../../../components/03-templates/FormSection";
 import PerustelutOpiskelijavuodet from "./Perustelut/PerustelutOpiskelijavuodet";
+import YhteenvetoLiitteet from "./Yhteenveto/YhteenvetoLiitteet";
+import PerustelutLiitteet from "./Perustelut/PerustelutLiitteet";
+import wizard from "../../../../../../i18n/definitions/wizard";
 
 const MuutospyyntoWizardPerustelut = ({
   changeObjects,
   intl,
+  intl: { formatMessage },
```
Comment (rapitkan) [inline src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardPerustelut.js]: There shouldn't be any need for this line. `intl` is already defined and it should contain the _formatMessage_ method.

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardTaloudelliset.js [hunk 5]
```diff
@@ -1,29 +1,26 @@
...
 import PropTypes from "prop-types";
 import * as R from "ramda";
 
 const MuutospyyntoWizardTaloudelliset = ({
   changeObjects,
   muutoshakemus,
   onChangesUpdate,
-  onStateUpdate
+  onStateUpdate,
+  intl: { formatMessage }
```
Comment (rapitkan) [inline src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardTaloudelliset.js]: There shouldn't be a need to define formatMessage. This should be fixed now or later.

File: src/services/lomakkeet/perustelut/koulutukset.js [hunk 6]
```diff
@@ -10,7 +10,16 @@ export const getKuljettajienJatkokoulutuslomake = (
     {
       anchor: code,
       styleClasses: ["px-10 py-10"],
-      title: "Jatkokoulutusta antavan koulutuskeskuksen tehtävä",
+      // title: "Jatkokoulutusta antavan koulutuskeskuksen tehtävä",
```
Comment (rapitkan) [inline src/services/lomakkeet/perustelut/koulutukset.js]: Commented out -> let's remove the line.

File: src/services/lomakkeet/perustelut/koulutukset.js [hunk 7]
```diff
@@ -43,7 +52,7 @@ export const getKuljettajienPeruskoulutuslomake = (
       styleClasses: ["px-10 py-10"],
       components: [
         {
-          anchor: "tehtavan-tarpeellisuus-title",
+          anchor: "paerustaso-title",
```
Comment (rapitkan) [inline src/services/lomakkeet/perustelut/koulutukset.js]: typo here?

File: src/services/lomakkeet/perustelut/koulutukset.js [hunk 8]
```diff
@@ -673,27 +702,19 @@ const johtaja = (orderCode, isReadOnly) => {
...
-                name: "StatusTextRow",
-                properties: {
-                  title: HAKEMUS_OTSIKOT.LIITE_OHJE.FI
-                }
-              }
-            ]
+            name: "StatusTextRow",
+            styleClasses: ["w-full"],
+            properties: {
+              title: HAKEMUS_OTSIKOT.LIITE_OHJE.FI
```
Comment (rapitkan) [inline src/services/lomakkeet/perustelut/koulutukset.js]: Let's fetch the localizations from our i18n based solution.

File: src/services/muutoshakemus/utils/changes-to-save.js [hunk 9]
```diff
@@ -168,6 +168,18 @@ export const getChangesToSave = (
...
+    //     const meta = stateObject.liitteet.categories[0].meta;
+    //     return {
+    //       kohde: meta.kohde, // object, tunniste=tutkinnotjakoulutukset
+    //       meta: {
+    //         changeObjects: [changeObj]
+    //       }
+    //       // tila: changeObj.properties.isChecked ? "LISAYS" : "POISTO",
+    //       // type: changeObj.properties.isChecked ? "addition" : "removal"
+    //     };
+    //   }, unhandledChangeObjects).filter(Boolean);
```
Comment (rapitkan) [inline src/services/muutoshakemus/utils/changes-to-save.js]: Commented out -> delete the lines.

File: src/services/muutospyynnot/muutospyyntoUtil.js [hunk 10]
```diff
@@ -397,3 +397,13 @@ export function setAttachmentUuids(
...
 }
+
+// Combine arrays recursively
+export function combineArrays(arrays) {
+  if (arrays && arrays.length > 0) {
+    const currentArray = arrays.pop();
+    if (currentArray) {
+      return currentArray.concat(combineArrays(arrays));
+    } else return [];
+  } else return [];
```
Comment (rapitkan) [inline src/services/muutospyynnot/muutospyyntoUtil.js]: `else` keywords aren't needed here, right? But it's ok like this.

Comment (sz1521) [reply src/services/muutospyynnot/muutospyyntoUtil.js]: Better to feel save than sorry :)

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js (hunk 3 above)
Comment (sz1521) [reply src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizard.js]: ->Refactoring later as ths also works.

File: src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardPerustelut.js (hunk 4 above)
Comment (sz1521) [reply src/scenes/Jarjestajat/Jarjestaja/Hakemukset/Muutospyynto/components/MuutospyyntoWizardPerustelut.js]: Some reason does not work without -> need to studied when all left localizations are done.

File: src/services/lomakkeet/perustelut/koulutukset.js (hunk 8 above)
Comment (sz1521) [reply src/services/lomakkeet/perustelut/koulutukset.js]: Need to done separative, injectIntl has problems with this structure also.

**Pull request merged**